src/bin/cargo/cli.rs

```diff
@@ -194,7 +196,27 @@ fn expand_aliases(
                 let new_args = cli()
                     .setting(AppSettings::NoBinaryName)
                     .get_matches_from_safe(alias)?;
-                let (expanded_args, _) = expand_aliases(config, new_args)?;
+
+                let (new_cmd, _) = new_args.subcommand();
+                if new_cmd == cmd {
+                    // Don't reexpand if the alias is a default-flags alias
+                    // eg. fmt = "fmt --verbose"
+                    return Ok((new_args, global_args));
+                }
+
+                already_expanded.push(cmd.to_string());
+                if already_expanded.contains(&new_cmd.to_string()) {
```

Comment:
happy to switch to a LinkedHashMap if we want cargo to support O(n*log(n)) resolution of dependent aliases instead of O(n^2). I figured n is probably small, and linked_hash_map crate is a new dependency - but happy to switch it if preferred.

Reply:
Why `LinkedHashMap` over [`indexmap`](https://docs.rs/indexmap/1.7.0/indexmap/map/struct.IndexMap.html)? But yes n is probably to small for it to matter.

Reply:
wasn't familiar with indexmap - looks great to me.
It still appears that it would be an additional dependency to cargo (not sure if you want to pull it in)

Reply:
I think this is fine for now, yeah, and if it's an issue there's a number of other data structures to use here instead.